infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/NamespaceProvisioner.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2012-2021 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.provision;
+
+import static org.eclipse.che.workspace.infrastructure.kubernetes.Constants.DEV_WORKSPACE_MOUNT_AS_ANNOTATION;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.Constants.DEV_WORKSPACE_MOUNT_LABEL;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.Constants.DEV_WORKSPACE_MOUNT_PATH_ANNOTATION;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.SecretBuilder;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import javax.inject.Inject;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.user.User;
+import org.eclipse.che.api.user.server.PreferenceManager;
+import org.eclipse.che.api.user.server.UserManager;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.NamespaceResolutionContext;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.api.shared.KubernetesNamespaceMeta;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * On namespace provisioning, creates k8s {@link Secret} profile and preferences from information
```

Comment:
Well, it feels a bit unusual that class is named NamespaceProvisioner but actually it's namespaceFactory who does it.

Taking into account that someday we may need other stuff as part of provisioning, what if we create some interface like `NamespaceConfigurator`, then namespaceFactory injects NamespaceConfigurators and call them each time when it provisions namespace.
And you have `UsersSecretsProvisioner implements NamespaceConfigurator`.

I know it may look like a bit overhead in this case, so feel free to just ignore that proposal.

Reply:
> Well, it feels a bit unusual that class is named NamespaceProvisioner but actually it's namespaceFactory who does it.

I kinda agree with this but also think that `NamespaceFactory` would deserve some refactoring to maybe take out the provisioning into separate class.

> Taking into account that someday we may need other stuff as part of provisioning, what if we create some interface like `NamespaceConfigurator`, then namespaceFactory injects NamespaceConfigurators and call them each time when it provisions namespace.
> And you have `UsersSecretsProvisioner implements NamespaceConfigurator`

I think its a good idea, but I would like to avoid injecting the `Configurators` into `NamespaceFactory` since its already too big imho.

If others see it as a useful change, I can refactor it. But if its not necessary, maybe changing the name `NamespaceProvisioner` to `NamespaceConfigurator` would be fine for now to avoid confusion.

Reply:
What if NamespaceConfigurators injected in NamespaceProvisioner ?

Reply:
That wouldn't change the fact that `NamespaceProvisioner` as is right now, doesn't actually do the provisioning right?

Reply:
ok. what do you propose?

Reply:
Basically just have a different name for `NamespaceProvisioner`.
So, how about we change `NamespaceProvisioner` into `NamespaceBuilder` and it will have injected different `XYZConfigurator` that implement `NamespaceConfigurator`. So in this case `UsersSecretsProvisioner implements NamespaceConfigurator`.

Service calls Builder -> Builder calls Factory.provision() and runs all injected Configurators.